credentials/credentials.go

```diff
@@ -161,17 +147,18 @@ func (c *tlsCreds) ClientHandshake(addr string, rawConn net.Conn, timeout time.D
 		cfg.ServerName = addr[:colonPos]
 	}
 	conn := tls.Client(rawConn, cfg)
-	if timeout == 0 {
-		err = conn.Handshake()
-	} else {
-		go func() {
-			errChannel <- conn.Handshake()
-		}()
-		err = <-errChannel
-	}
-	if err != nil {
-		rawConn.Close()
-		return nil, nil, err
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- conn.Handshake()
+	}()
+	select {
+	case err := <-errChannel:
+		if err != nil {
+			rawConn.Close()
+			return nil, nil, err
+		}
+	case <-ctx.Done():
```

Comment:
wait for go1.7 to resolve this issue? the main theme of this PR is to plumb the context in?